rfcs/2023-05-03-data-volume-metrics.md

```diff
@@ -0,0 +1,162 @@
+be emitted.
+
+If there is no `service` available, a service of `_no_service` will be emitted.
+
+#### `component_received_event_bytes_total`
+
+This metric is emitted by the framework [here][source_sender], so it looks like the only change needed is
+to add the service tag.
+
+#### `component_sent_event_bytes_total`
+
+For stream based sinks this will typically be the byte value returned by `DriverResponse::events_sent`.
+
+Despite being in the [Component Spec][component_spec], not all sinks currently conform to this.
+
+As an example, from a cursory glance over a couple of sinks:
+
+The Amqp sink currently emits this value as the length of the binary data that is sent. By the time the data has
+reached the code where the `component_sent_event_bytes_total` event is emitted, that event has been encoded
+and the actual estimated JSON size has been lost. The sink will need to be updated so that when the event is
+encoded, the encoded event together with the pre-encoded JSON bytesize will be sent to the service where the
+event is emitted.
+
+The Kafka sink also currently sends the binary size, but it looks like the estimated JSON bytesize is easily
+accessible at the point of emitting, so would not need too much of a change.
+
+To ensure that the correct metric is sent in a type-safe manner, we will wrap the estimated JSON size in a
+newtype:
+
+```rust
+pub struct JsonSize(usize);
+```
+
+The `EventsSent` metric will only accept this type.
+
+## Rationale
+
+The ability to visualize data flowing through Vector will allow users to ascertain
+the effectiveness of the current use of Vector. This will enable users to
+optimise their configurations to make the best use of Vector's features.
+
+## Drawbacks
+
+The additional tags being added to the metrics will increase the cardinality of
+those metrics if they are enabled.
+
+We will lose the ability to preregister the metrics since the tags will need to be
+dynamic. This will cause a noticable, but likely negligible performance loss.
+
+## Prior Art
+
+
+## Alternatives
+
+We could use an alternative metric instead of estimated JSON size.
+
+- *Network bytes* This provides a more accurate picture of the actual data being received
+  and sent by Vector, but will regularly produce different sizes for an incoming event
+  to an outgoing event.
+- *In memory size* The size of the event as held in memory. This may be more accurate in
+  determining the amount of memory Vector will be utilizing at any time, will often be
+  less accurate compared to the data being sent and received which is often JSON.
+
+## Outstanding Questions
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues after the RFC is approved:
+
+- [ ] Add the `source` field to the Event metadata to indicate the source the event has come from.
+- [ ] Update the Volume event metrics to take a `JsonSize` value. Use the compiler to ensure all metrics
```

Comment:
Does this step imply the audit of components?

Reply:
Yeah. Once the metrics take `JsonSize`, Vector will not compile until all the appropriate places have been updated.

Reply:
We could potentially also change the `EstimatedJsonEncodedSizeOf` trait to return a `JsonSize` to make that more obvious as well.